sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,535 @@
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.MongoClient;
+import com.mongodb.MongoClientURI;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.MongoDatabase;
+
+import org.bson.Document;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import javax.annotation.Nullable;
+
+/**
+ * IO to read and write data on MongoDB.
+ *
+ * <h3>Reading from MongoDB</h3>
+ *
+ * <p>MongoDbIO source returns a bounded collection of String as {@code PCollection<String>}.
+ * The String is the JSON form of the MongoDB Document.</p>
+ *
+ * <p>To configure the MongoDB source, you have to provide the connection URI, the database name
+ * and the collection name. The following example illustrates various options for configuring the
+ * source:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(MongoDbIO.read()
+ *   .withUri("mongodb://localhost:27017")
+ *   .withDatabase("my-database")
+ *   .withCollection("my-collection")
+ *   // above three are required configuration, returns PCollection<String>
+ *
+ *   // rest of the settings are optional
+ *
+ * }</pre>
+ *
+ * <p>The source also accepts an optional configuration: {@code withFilter()} allows you to
+ * define a JSON filter to get subset of data.</p>
+ *
+ * <h3>Writing to MongoDB</h3>
+ *
+ * <p>MongoDB sink supports writing of Document (as JSON String) in a MongoDB.</p>
+ *
+ * <p>To configure a MongoDB sink, you must specify a connection {@code URI}, a {@code Database}
+ * name, a {@code Collection} name. For instance:</p>
+ *
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...)
+ *   .apply(MongoDbIO.write()
+ *     .withUri("mongodb://localhost:27017")
+ *     .withDatabase("my-database")
+ *     .withCollection("my-collection")
+ *     .withNumberSplit(30)
+ *
+ * }</pre>
+ */
+// TODO instead of JSON String, does it make sense to populate the PCollection with BSON Document or
+//  DBObject ??
+public class MongoDbIO {
+
+  public static Read read() {
+    return new Read();
+  }
+
+  public static Write write() {
+    return new Write();
+  }
+
+  private MongoDbIO() {}
+
+  /**
+   * A {@link PTransform} to read data from MongoDB.
+   */
+  public static class Read extends PTransform<PBegin, PCollection<String>> {
+
+    public Read withUri(String uri) {
+      return new Read(uri, database, collection, filter, numberSplit);
+    }
+
+    public Read withDatabase(String database) {
+      return new Read(uri, database, collection, filter, numberSplit);
+    }
+
+    public Read withCollection(String collection) {
+      return new Read(uri, database, collection, filter, numberSplit);
+    }
+
+    public Read withFilter(String filter) {
+      return new Read(uri, database, collection, filter, numberSplit);
+    }
+
+    public Read withNumberSplit(int numberSplit) {
+      return new Read(uri, database, collection, filter, numberSplit);
+    }
+
+    protected String uri;
```

Comment:
Agreed for private and final. For the source, I just mimic what we do in other IOs like Kafka, or JMS.